eo-phi-normalizer/app/Main.hs

```diff
@@ -1,111 +1,222 @@
 {-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
 {-# LANGUAGE RecordWildCards #-}
+{-# OPTIONS_GHC -Wno-partial-fields #-}
 {-# OPTIONS_GHC -Wno-type-defaults #-}
 
-module Main where
+module Main (main) where
 
 import Control.Monad (unless, when)
 import Data.Foldable (forM_)
 
 import Control.Lens ((^.))
 import Data.Aeson (ToJSON)
-import Data.Aeson.Text (encodeToLazyText)
+import Data.Aeson.Encode.Pretty (Config (..), Indent (..), defConfig, encodePrettyToTextBuilder')
 import Data.List (nub)
+import Data.String.Interpolate (i)
+import Data.Text.Internal.Builder (toLazyText)
 import Data.Text.Lazy.Lens
-import Language.EO.Phi (Object (Formation), Program (Program), defaultMain, parseProgram, printTree)
-import Language.EO.Phi.Rules.Common (Context (..), applyRules, applyRulesChain)
+import GHC.Generics (Generic)
+import Language.EO.Phi (Object (Formation), Program (Program), parseProgram, printTree)
+import Language.EO.Phi.Metrics.Collect (collectMetrics)
+import Language.EO.Phi.Rules.Common (applyRules, applyRulesChain)
+import Language.EO.Phi.Rules.Common qualified as Common
 import Language.EO.Phi.Rules.Yaml (RuleSet (rules, title), convertRule, parseRuleSetFromFile)
-import Options.Generic
-import System.IO (IOMode (WriteMode), hClose, hPutStr, hPutStrLn, openFile, stdout)
+import Options.Applicative
+import Options.Applicative.Types qualified as Optparse (Context (..))
+import System.IO (IOMode (WriteMode), getContents', hFlush, hPutStr, hPutStrLn, openFile, stdout)
 
-data CLINamedParams = CLINamedParams
+data CLI'TransformPhi = CLI'TransformPhi
   { chain :: Bool
-  , rulesYaml :: Maybe String
-  , outPath :: Maybe String
+  , rulesPath :: String
+  , outputFile :: Maybe String
   , single :: Bool
   , json :: Bool
+  , inputFile :: Maybe FilePath
+  , program :: Maybe String
   }
-  deriving (Generic, Show, ParseRecord, Read, ParseField)
+  deriving (Show)
 
-instance ParseFields CLINamedParams where
-  parseFields _ _ _ _ =
-    CLINamedParams
-      <$> parseFields (Just "Print out steps of reduction") (Just "chain") (Just 'c') Nothing
-      <*> parseFields (Just "Path to the Yaml file with custom rules") (Just "rules-yaml") Nothing Nothing
-      <*> parseFields (Just "Output file path (defaults to stdout)") (Just "output") (Just 'o') Nothing
-      <*> parseFields (Just "Print a single normalized expression") (Just "single") (Just 's') Nothing
-      <*> parseFields (Just "Print JSON") (Just "json") (Just 'j') Nothing
+data CLI'MetricsPhi = CLI'MetricsPhi
+  { inputFile :: Maybe FilePath
+  , outputFile :: Maybe FilePath
+  , program :: Maybe String
+  }
+  deriving (Show)
+
+data CLI
+  = CLI'TransformPhi' CLI'TransformPhi
+  | CLI'MetricsPhi' CLI'MetricsPhi
+  deriving (Show)
+
+_FILE :: String
+_FILE = "FILE"
+
+_PROGRAM :: String
+_PROGRAM = "PROGRAM"
+
+fileMetavar :: Mod OptionFields a
+fileMetavar = metavar _FILE
+
+inputFileLongName :: String
+inputFileLongName = "input-file"
```

Comment:
What is the reason behind these constants? And why is their naming (or casing) irregular?

Reply:
> What is the reason behind these constants?

I reuse them.

> And why is their naming (or casing) irregular?
- uppercase are for metavar names.
- camelcase are for option names

I can make them camelcase if you'd like.